torchtune/modules/loss/cross_entropy_loss.py

```diff
@@ -143,3 +143,142 @@ def forward(
+    >>> loss = LigerLinearCrossEntropy(...)
+    >>> loss.set_model_output(model)
+    >>> loss.apply_compile_strategy()
+    """
+
+    def __init__(self, ignore_index: int = -100):
+        super().__init__()
+        try:
+            import liger_kernel.ops.fused_linear_cross_entropy
+
+            self.fused_linear_ce = liger_kernel.ops.fused_linear_cross_entropy
+        except ImportError as err:
+            raise ImportError(
+                "liger_kernel is required for LigerLinearCrossEntropy but not installed. "
+                "Please install it before using this loss function."
+            ) from err
+        self.linear_projection = None
+        self.ignore_index = ignore_index
+
+    def set_model_output(self, model: nn.Module) -> None:
+        """Modify model output to match the expected input for the loss function.
+
+        Args:
+            model (nn.Module): The model whose output layer will be used for the loss computation.
+
+        Raises:
+            ValueError: If model.output doesn't have required weight and bias parameters
+        """
+        model.skip_output_layer = True
+        self.linear_projection = model.output
+
+        # Validate the projection layer has required parameters
+        if not hasattr(self.linear_projection, "weight"):
+            raise ValueError(
+                "Model output layer must have a weight parameter for LigerLinearCrossEntropy"
+            )
+
+    def apply_compile_strategy(self, *args, **kwargs):
+        """Triton kernels are already JIT-compiled so no additional compilation needed."""
+        return self
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        targets: torch.Tensor,
+    ) -> torch.Tensor:
+        """Compute the fused linear cross entropy loss.
+
+        Args:
+            hidden_states (torch.Tensor): Hidden state of the model, pre projection. Shape ``[bsz, seq_len, emb_dim]``
+            targets (torch.Tensor): Labels for the model. Shape ``[bsz, seq_len]``
+
+        Returns:
+            torch.Tensor: The computed loss value
+
+        Raises:
+            RuntimeError: If set_model_output() was not called before forward
+        """
+        if self.linear_projection is None:
+            raise RuntimeError("Must call set_model_output() before forward()")
+        if isinstance(hidden_states, DTensor):
+            hidden_states = hidden_states.full_tensor()
+
+        batch_size, seq_len, emb_dim = hidden_states.shape
+        hidden_states = hidden_states.reshape(
+            -1, emb_dim
+        )  # [batch_size*seq_len, emb_dim]
+        targets = targets.reshape(-1)  # [batch_size*seq_len]
+
+        mask = targets != self.ignore_index
+        total_elements = mask.sum()
+
+        w = self.linear_projection.weight
+        if isinstance(w, DTensor):
+            mesh, placements = w.device_mesh, w.placements
+            w = w.full_tensor()
+            if not hasattr(self, "_w_hook_registered"):
```

Comment:
I think "full_tensor" handles gradient placement and we don't need to do this ourselves [link](https://github.com/pytorch/pytorch/blob/35ddad284d350681073aa06a33904d1f31e6fff0/torch/distributed/tensor/_api.py#L571). I can test removing this though.